Plot from txt (#13)
MilanoBicocca-pix/BeamspotTools: quickplot_instructions.md

# Instruction to produce a new beamspot payload 

These instruction explains how to fit a new beamspot in a given run, alignment and release, and how to create a new payload from the fit results for a new global tag. An example is described at the end.

## work space setup

Download your favourite CMSSW release, e.g. *13_2_0_pre3*

```bash
cmsrel CMSSW_13_2_0_pre3
cd CMSSW_13_2_0_pre3/src
cmsenv
```

Fetch the beamspot tools and compile

```bash
git-cms-addpkg RecoVertex/BeamSpotProducer
cd RecoVertex/BeamSpotProducer/python
git clone https://github.com/MilanoBicocca-pix/BeamspotTools.git
cd $CMSSW_BASE/src
scram b -r -j8
```

The *uncertainty* python module is used in some scripts. It should be included in recent CMSSW releases, if not:

```bash
pip install --user uncertainties
```

## beamspot fit

The beamspot fit is configured by the [BeamFit_custom_workflow.py](test/BeamFit_custom_workflow.py) file, which is an extension  of the standard *BeamFit_\*_Workflow.py* configuration files. To run the fit:

```bash
cmsRun BeamFit_custom_workflow.py     \
  jobName="name"                        \

>   globalTag="tag"                       \

Reviewer: ```suggestion
  globalTag="Global Tag name"                       \
```

  refit=bool                            \
  runs="run:ls-run:ls"                  \
  dataset="/dataset/period/tier"        \
  storage="root://service"              \
  highPurity=bool                       \
  tracks="collection_name"              \

>   alignment="record1:tag1,record2:tag2" \

Reviewer: ```suggestion
  alignment="record1:tag1:label1,record2:tag2:label2" \
```

*<sup>(\*\*)</sup>&nbsp;the argument or --inputFiles is mandatory*  

The result of the fit will be a *.txt* file containing the list of runs and the fit results in some format **not** compatible with the CMS database (see below how to convert this file to a *.db* file).  
If the *saveRootFile* argument is set to *true*, a *.root* file will be created containing the fit results. Both the *.txt* and *.root* file name are set by the *jobName* argument.   
A file with the "_filelist.txt" prefix is also creted. It contains the list of files used in the fit (**NOTE**: if specific lumisections are analyzed by setting the *runs* argument, the fitter might consume more files than needed as there is no way to fetch single lumisections from DAS. These files will also appear in this list but their content will be skipped).  
(NOT YET) A file with the "_cfg.py" prefix will be written and will contain the full configuration of the job.
  
**NOTE** the track collection from *ALCARECO* is *ALCARECOTkAlMinBias*, for *RAW* the label is *generalTracks*.  
**NOTE** Other parameters of the fitter are present in the configuration file and usually there is no need to modify them.  

> **NOTE** by default, the fit is run on five lumisections. This can be changed by editing the parameters (note: set them equal).  

Reviewer: ```suggestion
**NOTE** by default, the fit is run on one lumisection. This can be changed by editing the parameters (note: set them equal).  
```

**NOTE** if a track refit is done and the alignment is not changed, the beamspot result will still be different. This is because, in such case, the configuration introduces additional cuts on the PVs.
**NOTE** the fitter appends the results of the fit to the *.txt* file live, thread safety is **NOT** guaranteed.

## beamspot plot

> **NOTE** the plotting script is a work in progress. At the moment it can plot a single file in a format good for check-ups (bs parameter vs. time). For instructions on how to produce complete BS plots, see [here]().  

Reviewer: ```suggestion
**NOTE** the plotting script is a work in progress. At the moment it can plot a single file in a format good for check-ups (bs parameter vs. time).
```

**NOTE** this script does not work with any CMSSW release. Python 3 is required. All the modules used should be available in the default python3 distribution, except [ROOT](https://root.cern/).  
**NOTE** [auth-get-sso-cookie](https://gitlab.cern.ch/authzsvc/tools/auth-get-sso-cookie) is required. This is available by default on lxplus or can be installed user-side following the instructions of the repo.  
**NOTE** a valid kerberos ticket is required to download data from OMS.

To plot the results, run [plotFromTxt.py](plot/plotFromTxt.py):

```bash

> python3 plotFromTxt.py --input input_file_list --output output_dir --streams 5

Reviewer: ```suggestion
python3 plotFromTxt.py --input input_file_list --output output_dir --streams 5 [--canfail]
```

## example

We want to fit the beamspot on ALCARECO files for run 370772 of 2023D, updating the tracker alignemtn tag to *TrackerAlignment_collisions23_forHLT_v5*. We will use a single local file from the */StreamExpress/Run2023D-TkAlMinBias-Express-v2/ALCARECO* dataset (to run on the full run, use *--dataset=dataset_name* instead of *--inputFiles*)

```bash
xrdcp root://cms-xrd-global.cern.ch//store/express/Run2023D/StreamExpress/ALCARECO/TkAlMinBias-Express-v2/000/370/772/00001/63cbf285-ce54-4ad2-bc9b-833799330067.root AlcaFile_2023D.root
cmsRun BeamFit_custom_workflow.py         \
  jobName="localTest_refit_alignmentUpdate" \
  inputFiles="file:AlcaFile_2023D.root"     \
  runs="370772:min-370772:max"              \
  globalTag="130X_dataRun3_Express_v3"      \
  tracks=ALCARECOTkAlMinBias                \
  refit=True                                \
  updateGT="TrackerAlignmentRcd:TrackerAlignment_collisions23_forHLT_v5:"
```
The result can be compared to running without the *--updateGT* argument to see the effect of the new alignment.  
  
To plot and check the result:

```bash

> python3 plotFromTxt.py --input --output Run2023_370772 --streams 10

Reviewer: ```suggestion
python3 plotFromTxt.py --input localTest_refit_alignmentUpdate.txt --output Run2023_370772 --streams 10
```
